core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java

```diff
@@ -0,0 +1,170 @@
+  private final Configuration conf;
+  private String warehouseUri;
+
+  public HadoopCatalog(Configuration conf, String warehouseUri) {
+    this.conf = conf;
+
+    if (warehouseUri != null) {
+      this.warehouseUri = warehouseUri;
+    } else {
+      String fsRoot = conf.get("fs.defaultFS");
+      Path warehousePath = new Path(fsRoot, ICEBERG_HADOOP_WAREHOUSE_BASE);
+      try {
+        FileSystem fs = Util.getFs(warehousePath, conf);
+        if (!fs.isDirectory(warehousePath)) {
+          if (!fs.mkdirs(warehousePath)) {
+            throw new IOException("failed to create warehouse for hadoop catalog");
+          }
+        }
+        this.warehouseUri = fsRoot + "/" + ICEBERG_HADOOP_WAREHOUSE_BASE;
+      } catch (IOException e) {
+        throw new RuntimeIOException("failed to create directory for warehouse", e);
+      }
+    }
+  }
+
+  public HadoopCatalog(Configuration conf) {
+    this(conf, null);
+  }
+
+  @Override
+  public org.apache.iceberg.Table createTable(
+      TableIdentifier identifier, Schema schema, PartitionSpec spec, Map<String, String> properties) {
+    Preconditions.checkArgument(identifier.namespace().levels().length >= 1,
+            "Missing database in table identifier: %s", identifier);
+    Path tablePath = new Path(defaultWarehouseLocation(identifier));
+    try {
+      FileSystem fs = Util.getFs(tablePath, conf);
+      if (!fs.isDirectory(tablePath)) {
+        fs.mkdirs(tablePath);
+      } else {
+        throw new AlreadyExistsException("the directory %s for table %s is already exists.",
+          tablePath.toString(), identifier.toString());
+      }
+    } catch (IOException e) {
+      throw new RuntimeIOException("failed to create directory", e);
+    }
+    return super.createTable(identifier, schema, spec, null, properties);
+  }
+
+  public org.apache.iceberg.Table createTable(
+          TableIdentifier identifier, Schema schema, PartitionSpec spec) {
+    Preconditions.checkArgument(identifier.namespace().levels().length >= 1,
+            "Missing database in table identifier: %s", identifier);
+    return createTable(identifier, schema, spec, null, null);
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier identifier) {
+    Preconditions.checkArgument(identifier.namespace().levels().length >= 1,
+                "Missing database in table identifier: %s", identifier);
+    return new HadoopTableOperations(new Path(defaultWarehouseLocation(identifier)), conf);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier tableIdentifier) {
+    String[] levels = tableIdentifier.namespace().levels();
+    String tableName = tableIdentifier.name();
+    
+    List<String> path = new ArrayList<>();
+    path.add(warehouseUri);
+    path.add(levels[0] + ".db");
+    for (int i = 1; i < levels.length; i++){
+      path.add(levels[i]);
+    }
+    path.add(tableName);
+
+    return Joiner.on("/").join(path.iterator());
```

Comment:
I think it would be simpler to use a StringBuilder instead of a List and a joiner:

```java
StringBuilder sb = new StringBuilder();
sb.append(warehouseLocation).append('/');
for (String level : tableIdentifier.namespace().levels()) {
  sb.append(level).append('/');
}
sb.append(tableName);
return sb.toString;
```